#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-58_volume.dpatch by clausmuus at vdrportal
## http://vdr-portal.de/board/thread.php?threadid=96574
##
##
## All lines beginning with `## DP:' are a description of the patch.

@DPATCH@
diff -urNad vdr-1.7.14/config.c vdr-1.7.14~/config.c
--- vdr-1.7.14/config.c	2010-06-05 10:48:35.165440727 +0200
+++ vdr-1.7.14~/config.c	2010-06-05 10:57:00.435938055 +0200
@@ -408,6 +408,8 @@
   ReloadMarks = 0;
   CurrentChannel = -1;
   CurrentVolume = MAXVOLUME;
+  VolumeSteps = 51;
+  VolumeLinearize = 0;
   CurrentDolby = 0;
   InitialChannel = 0;
   InitialVolume = -1;
@@ -632,6 +634,8 @@
   else if (!strcasecmp(Name, "CurrentDolby"))        CurrentDolby       = atoi(Value);
   else if (!strcasecmp(Name, "InitialChannel"))      InitialChannel     = atoi(Value);
   else if (!strcasecmp(Name, "InitialVolume"))       InitialVolume      = atoi(Value);
+  else if (!strcasecmp(Name, "VolumeSteps"))         VolumeSteps        = atoi(Value);
+  else if (!strcasecmp(Name, "VolumeLinearize"))     VolumeLinearize    = atoi(Value);
   else if (!strcasecmp(Name, "ChannelsWrap"))        ChannelsWrap       = atoi(Value);
   else if (!strcasecmp(Name, "EmergencyExit"))       EmergencyExit      = atoi(Value);
   else if (!strcasecmp(Name, "ShowRecDate"))         ShowRecDate        = atoi(Value);
@@ -759,6 +763,8 @@
   Store("CurrentDolby",       CurrentDolby);
   Store("InitialChannel",     InitialChannel);
   Store("InitialVolume",      InitialVolume);
+  Store("VolumeSteps",        VolumeSteps);
+  Store("VolumeLinearize",    VolumeLinearize);
   Store("ChannelsWrap",       ChannelsWrap);
   Store("EmergencyExit",      EmergencyExit);
   Store("ShowRecDate",        ShowRecDate);
diff -urNad vdr-1.7.14/config.h vdr-1.7.14~/config.h
--- vdr-1.7.14/config.h	2010-06-05 10:48:35.165440727 +0200
+++ vdr-1.7.14~/config.h	2010-06-05 10:55:04.183949188 +0200
@@ -304,6 +304,8 @@
   int ReloadMarks;
   int CurrentChannel;
   int CurrentVolume;
+  int VolumeSteps;
+  int VolumeLinearize;
   int CurrentDolby;
   int InitialChannel;
   int InitialVolume;
diff -urNad vdr-1.7.14/device.c vdr-1.7.14~/device.c
--- vdr-1.7.14/device.c	2010-06-05 10:48:35.169439731 +0200
+++ vdr-1.7.14~/device.c	2010-06-05 10:54:16.252020820 +0200
@@ -11,6 +11,7 @@
 #include <errno.h>
 #include <sys/ioctl.h>
 #include <sys/mman.h>
+#include <math.h>
 #include "audio.h"
 #include "channels.h"
 #include "i18n.h"
@@ -966,8 +967,9 @@
 void cDevice::SetVolume(int Volume, bool Absolute)
 {
   int OldVolume = volume;
-  volume = min(max(Absolute ? Volume : volume + Volume, 0), MAXVOLUME);
-  SetVolumeDevice(volume);
+  float VolumeDelta = (float)MAXVOLUME/Setup.VolumeSteps;
+  volume = min(max((int)(floor((Absolute ? Volume : volume + Volume)/VolumeDelta+0.5)*VolumeDelta), 0), MAXVOLUME);
+  SetVolumeDevice((int)(pow((float)volume/MAXVOLUME, Setup.VolumeLinearize>0 ? ((float)Setup.VolumeLinearize/10+1) : (1/((float)Setup.VolumeLinearize/-10+1)))*MAXVOLUME));
   Absolute |= mute;
   cStatus::MsgSetVolume(Absolute ? volume : volume - OldVolume, Absolute);
   if (volume > 0) {
diff -urNad vdr-1.7.14/device.h vdr-1.7.14~/device.h
--- vdr-1.7.14/device.h	2010-06-05 10:48:35.169439731 +0200
+++ vdr-1.7.14~/device.h	2010-06-05 10:52:32.489438716 +0200
@@ -31,7 +31,7 @@
 #define MAXPIDHANDLES      64 // the maximum number of different PIDs per device
 #define MAXRECEIVERS       16 // the maximum number of receivers per device
 #define MAXVOLUME         255
-#define VOLUMEDELTA         5 // used to increase/decrease the volume
+#define VOLUMEDELTA       (MAXVOLUME/Setup.VolumeSteps) // used to increase/decrease the volume 
 
 enum eSetChannelResult { scrOk, scrNotAvailable, scrNoTransfer, scrFailed };
 
diff -urNad vdr-1.7.14/menu.c vdr-1.7.14~/menu.c
--- vdr-1.7.14/menu.c	2010-06-05 10:48:35.173440117 +0200
+++ vdr-1.7.14~/menu.c	2010-06-05 10:58:05.943945407 +0200
@@ -3546,6 +3546,8 @@
   Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$Channel entry timeout (ms)"), &data.ChannelEntryTimeout, 0));
   Add(new cMenuEditChanItem(tr("Setup.Miscellaneous$Initial channel"),            &data.InitialChannel, tr("Setup.Miscellaneous$as before")));
   Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$Initial volume"),             &data.InitialVolume, -1, 255, tr("Setup.Miscellaneous$as before")));
+  Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$Volume steps"),               &data.VolumeSteps, 5, 255));
+  Add(new cMenuEditIntItem( tr("Setup.Miscellaneous$Volume linearize"),           &data.VolumeLinearize, -20, 20));
   Add(new cMenuEditBoolItem(tr("Setup.Miscellaneous$Channels wrap"),              &data.ChannelsWrap));
   Add(new cMenuEditBoolItem(tr("Setup.Miscellaneous$Emergency exit"),             &data.EmergencyExit));
 }
